functorch/_src/aot_autograd.py

```diff
@@ -1066,25 +1074,145 @@ def format_guard_bug_msg(aot_config, expected):
+#               y.transpose_(0, 2)
+#
+#           x = torch.randn(2, 3, 4)
+#           f(x, x)
+#
+#       The ordering of the transposes inside f dictates whether or not
+#       you get [4, 2, 3] or [3, 4, 2].  This means that you cannot precompute
+#       what metadata mutations should get applied to each input; you need to
+#       assume they aren't duplicates (what we do today) or preserve
+#       the original metadata mutations exactly in order, so that they work
+#       for any duplicate configuration.
+#
+#       CON: It does not (naively) work if you mutate the data of inputs.
+#       In particular, leaf tensors that require grad cannot be mutated,
+#       this makes it impossible to differentiate with respect to the original
+#       base.
+#
+# 2. For every duplicate argument to the function, remove it, so it is
+#    no longer part of the "true" signature:
+#
+#       PRO: Implemented naively, it still works for metadata/data mutation.
+#
+#       CON: The resulting compiled graph is duplicate-specialized: it only
+#       works if future calls duplicate arguments in exactly the same way.
+#       Horribly, Dynamo doesn't guard on this at the moment.  But even if
+#       it did, you could still end up recompiling a bunch of each duplicate.
+#
+# Our strategy is to do (1) if we can, and do (2) otherwise, erroring if
+# Dynamo's guards are not enough.  In practice, this seems to cover
+# everything.
+#
+def aot_wrapper_dedupe(flat_fn, flat_args: List[Tensor], aot_config: AOTConfig, *, compiler_fn):
+    # Get information about whether or not flat_fn mutates its arguments
+    # or not
+    try:
+        with enable_python_dispatcher():
+            fw_metadata, _out, _num_aliasing_metadata_outs = run_functionalized_fw_and_collect_metadata(
+                flat_fn
+            )(*flat_args)
+    except RuntimeError as e:
+        logging.warning(
+            "Failed to collect metadata on function, produced code may be suboptimal.  "
+            "Known situations this can occur are inference mode only compilation involving "
+            "resize_ or prims (!schema.hasAnyAliasInfo() INTERNAL ASSERT FAILED); "
+            "if your situation looks different please file a bug to PyTorch.",
+            exc_info=True
+        )
+        # Analysis failed, fall back to duplicate specialize
+        # TODO: Known analysis problems:
+        #   - resize_: TestInductorOpInfoCPU.test_comprehensive_resize__cpu_bool
+        #   - prims: test_tmp_not_defined_issue1_cpu
+        pass
+    else:
+        # Strategy 1: For any input that is not mutated, we can leafify it if we
+        # need to remove a duplicate.
+        leaf_flat_args = []
+        args_set = set()
+        ok = True
+
+        for i, a in enumerate(flat_args):
+            if a not in args_set:
+                args_set.add(a)
+                leaf_flat_args.append(a)
+            elif fw_metadata.mutated_input_info[i] == MutationType.none:
```

Comment:
Right now, won't we fallback to strategy #2 sometimes when we don't have to? It looks like when you have dupe args, it always uses the first instance directly, and tries to detach() all later instances. If you have:
```
def f(x, y):
    y.mutate_(...)
```

Then we could use in theory use strategy #1 by passing in `f(x.detach(), x)`.

If you think it's not worth beefing up strategy #1 today, that seems fine to me. AKA the bad situation would be we use strategy #2 when we could have used strategy #1, and strategy #2 ends up assert failing due to dynamo not producing enough guards. But we want to fix dynamo's guards later anyway, IIUC.

Reply:
Your proposal here does not work because you may have:

```
def f(x, y):
    y.unsqueeze_(2)
    x.blah()
```

and if x is y, then the metadata mutation must reflect to x. Your alternate detach strategy still does not propagate the metadata.

Reply:
whoops. Ok, different direction: If we're given:
```
def f(x, y):
    x.unsqueeze_(2)
    y.blah()
...
f(x, x)
```

Then we want to skip strategy 1 here, but won't the code above do strategy 1? Right now it just checks if the dupe arg has any mutations, but we also need to check if the original dupe (that was already added to `args_set`) also has any mutations.

Reply:
Yes, you're right!

Reply:
Actually, it is fine! The reason is that when we run mutation detection, we do so mimicking the aliasing we were called with. So x is y when x.unsqueeze_(2) is called, and then in the end we say that BOTH x and y had their metadata mutated (even though there was only one metadata mutation call.) I wrote a test for this though which I'll commit in.